### PR TITLE
MM-13606 Remove consumeAndClose and clean up integration response handling

### PR DESCRIPTION
**Pull request by hmhealey**
It may be easier to review this PR if you look at the individual commits since it accomplishes two main things:
1. I removed the `consumeAndClose` function used throughout the app layer since it seems like you don't need to read from a TCP request's body any more to have the connection reused
2. I did some refactoring of the outgoing webhook and slash command code so that I add some tests for that code

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-13606

#### Checklist
- Added or updated unit tests (required for all new features)
- Touches critical sections of the codebase (auth, upgrade, etc.)

### CODE REVIEW

File: app/command.go [hunk 1]
```diff
@@ -202,98 +223,112 @@ func (a *App) ExecuteCommand(args *model.CommandArgs) (*model.CommandResponse, *
...
-			}
+	return a.doCommandRequest(cmd, p)
+}
 
-			response.TriggerId = clientTriggerId
+func (a *App) doCommandRequest(cmd *model.Command, p url.Values) (*model.Command, *model.CommandResponse, *model.AppError) {
+	// Prepare the request
+	var req *http.Request
+	if cmd.Method == model.COMMAND_METHOD_GET {
+		req, _ = http.NewRequest(http.MethodGet, cmd.URL, nil)
```
Comment (levb) [inline app/command.go]: I _just_ fixed a bug on this. We must not ignore error codes returned from standard library.

File: app/command.go [hunk 2]
```diff
@@ -176,24 +176,45 @@ func (a *App) ExecuteCommand(args *model.CommandArgs) (*model.CommandResponse, *
 
 	args.TriggerId = triggerId
 
-	if provider != nil {
-		if cmd := provider.GetCommand(a, args.T); cmd != nil {
-			response := provider.DoCommand(a, args, message)
-			return a.HandleCommandResponse(cmd, args, response, true)
-		}
+	if cmd, response := a.tryExecuteBuiltInCommand(args, trigger, message); cmd != nil && response != nil {
```
Comment (levb) [inline app/command.go]: 2/5 I'd suggest the names `a.executeBuiltInCommand` and `a.executeCustomCommand` for even though `try...` is a little more intention-revealing. I think the code clarity would be improved by consistent naming altogether with ExecutePluginCommand.

File: app/command.go [hunk 3]
```diff
@@ -176,24 +176,45 @@ func (a *App) ExecuteCommand(args *model.CommandArgs) (*model.CommandResponse, *
...
+		return a.HandleCommandResponse(cmd, args, response, true)
 	}
 
-	cmd, response, appErr := a.ExecutePluginCommand(args)
-	if appErr != nil {
+	if cmd, response, appErr := a.ExecutePluginCommand(args); appErr != nil {
```
Comment (levb) [inline app/command.go]: I am not sure if we have a preferred style recommendation for "horizontal" vs "vertical" error handling. I have a personal preference for the latter, also consistent with https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow, but most important I think we ought to be consistent.

File: app/command.go [hunk 4]
```diff
@@ -202,98 +223,112 @@ func (a *App) ExecuteCommand(args *model.CommandArgs) (*model.CommandResponse, *
...
+	// Send the request
+	resp, err := a.HTTPService.MakeClient(false).Do(req)
+	if err != nil {
+		return cmd, nil, model.NewAppError("command", "api.command.execute_command.failed.app_error", map[string]interface{}{"Trigger": cmd.Trigger}, err.Error(), http.StatusInternalServerError)
+	}
+
+	defer resp.Body.Close()
+
+	// Handle the response
+	body := io.LimitReader(resp.Body, MaxIntegrationResponseSize)
```
Comment (levb) [inline app/command.go]: See https://mattermost.atlassian.net/browse/MM-13519, FYI

File: app/command.go (hunk 2 above)
Comment (hmhealey) [reply app/command.go]: Will do

File: app/command.go (hunk 3 above)
Comment (hmhealey) [reply app/command.go]: I can change that. I forgot we had a recent push to change it to be that way, so I think I've actually reverted some recent work to do that. Do you have any preference on whether the if two statements would be separate as before or if they would use an else if? I kind of like the logical grouping provided by else if

File: app/command.go (hunk 1 above)
Comment (hmhealey) [reply app/command.go]: Yep, I'll add that in

File: app/command.go (hunk 2 above)
Comment (hmhealey) [reply app/command.go]: Actually, `ExecutePluginCommand` isn't exposed publicly anywhere, so I'll just rename that to match

File: app/webhook.go [hunk 5]
```diff
@@ -101,55 +103,66 @@ func (a *App) TriggerWebhook(payload *model.OutgoingWebhookPayload, hook *model.
...
+					mlog.Error(fmt.Sprintf("Failed to create response post, err=%v", err))
 				}
 			}
-		}(url))
+		})
 	}
 }
 
+func (a *App) doOutgoingWebhookRequest(url string, body io.Reader, contentType string) (*model.OutgoingWebhookResponse, error) {
+	req, _ := http.NewRequest("POST", url, body)
```
Comment (levb) [inline app/webhook.go]: Is this one going to be fixed? I am working on https://mattermost.atlassian.net/browse/MM-13659, trying to cover all other cases

File: app/command.go [hunk 6]
```diff
@@ -202,98 +230,119 @@ func (a *App) ExecuteCommand(args *model.CommandArgs) (*model.CommandResponse, *
...
+
+	if resp.StatusCode != http.StatusOK {
+		bodyBytes, _ := ioutil.ReadAll(body)
```
Comment (crspeller) [inline app/command.go]: Error ignored.

File: app/oauth.go [hunk 7]
```diff
@@ -820,25 +817,26 @@ func (a *App) AuthorizeOAuthUser(w http.ResponseWriter, r *http.Request, service
...
-		return nil, "", stateProps, model.NewAppError("AuthorizeOAuthUser", "api.user.authorize_oauth_user.token_failed.app_error", nil, readErr.Error(), http.StatusInternalServerError)
-	}
+	resp, err = a.HTTPService.MakeClient(true).Do(req)
+	if err != nil {
+		return nil, "", stateProps, model.NewAppError("AuthorizeOAuthUser", "api.user.authorize_oauth_user.service.app_error", map[string]interface{}{"Service": service}, err.Error(), http.StatusInternalServerError)
+	} else if resp.StatusCode != http.StatusOK {
+		defer resp.Body.Close()
 
-	if resp.StatusCode != http.StatusOK {
+		bodyBytes, _ := ioutil.ReadAll(resp.Body)
```
Comment (crspeller) [inline app/oauth.go]: Error ignored.

File: app/command.go (hunk 6 above)
Comment (hmhealey) [reply app/command.go]: We can't really handle those errors any better, but I added comments explaining why we're okay ignoring them

**Comment by hmhealey** [comment]
I'm not cherry picking this immediately since we want to test that community-daily performs fine with consumeAndClose removed before doing that

**Pull request merged**